java/java.lsp.server/vscode/src/jdk/validation/project.ts

```diff
@@ -0,0 +1,171 @@
+import * as xml2js from 'xml2js';
+import * as fs from 'fs';
+import * as path from 'path';
+import * as vscode from 'vscode';
+import * as os from 'os';
+import { normalizeJavaVersion } from './javaUtil';
+
+const GET_PROJECT_INFO = 'nbls.project.info';
+const GRADLE_TARGET_COMPATIBILITY_REGEX = /targetCompatibility\s*=\s*(?:JavaVersion\.toVersion\(['"](\d+(\.\d+)?)['"]\)|['"](\d+(\.\d+)?)['"])/
+const GRADLE_SOURCE_COMPATIBILITY_REGEX = /sourceCompatibility\s*=\s*(?:JavaVersion\.toVersion\(['"](\d+(\.\d+)?)['"]\)|['"](\d+(\.\d+)?)['"])/
+
+export enum BuildSystemType {
+    MAVEN = 'Maven',
+    GRADLE = 'Gradle',
+    UNKNOWN = 'Unknown'
+}
+
+export async function getProjectFrom(projectUri: vscode.Uri): Promise<Project | undefined> {
+    const projectInfos: any[] = await vscode.commands.executeCommand(GET_PROJECT_INFO, projectUri.toString(), { projectStructure: true });
```

Comment:
If searching for maximum required java, add `recursive: true` to also cover grand-child subprojects.